age.md

```diff
@@ -0,0 +1,296 @@
+Curve25519 base point from RFC 7748, Section 4.1.
+
+The recipient is encoded as Bech32 with HRP `age`.
+
+    age1zvkyg2lqzraa2lnjvqej32nkuu0ues2s82hzrye869xeexvn73equnujwj
+
+Note that Bech32 strings can only be all uppercase or all lowercase, but the
+checksum is always computed over the lowercase string.
+
+#### X25519 recipient stanza
+
+An X25519 recipient stanza has two arguments.
+
+    -> X25519 O6DLx/wDIawpUC978NSPjYvrfDtJVnZApXKp4FMPHCY
+    3aKjozt9agh7jGmvOKvR4iax41Wl4zj95MKK4X9JuWc
+
+The first is the fixed string `X25519` and the second is the base64-encoded
+ephemeral share computed by the recipient implementation as follows:
+
+    ephemeral secret = read(CSPRNG, 32)
+    ephemeral share = X25519(ephemeral secret, basepoint)
+
+A new ephemeral secret MUST be generated for each stanza and each file.
+
+The body of the recipient stanza is computed by the recipient implementation as
+
+    salt = ephemeral share || recipient
+    info = "age-encryption.org/v1/X25519"
+    shared secret = X25519(ephemeral secret, recipient)
+    wrap key = HKDF-SHA-256(ikm = shared secret, salt, info)
+    body = ChaCha20-Poly1305(key = wrap key, plaintext = file key)
+
+where the ChaCha20-Poly1305 nonce is fixed as 12 0x00 bytes.
+
+The identity implementation MUST ignore any stanza that does not have `X25519`
+as the first argument, and MUST otherwise reject any stanza that has more or
+less than two arguments, or where the second argument is not a canonical
+encoding of a 32-byte value. It MUST check that the body length is exactly 32
+bytes before attempting to decrypt it.
+
+The identity implementation computes the shared secret as follows:
+
+    shared secret = X25519(identity, ephemeral share)
+
+If the shared secret is all 0x00 bytes, the identity implementation MUST abort.
+
+Finally, it derives the key as above and decrypts the file key in the body.
+
+### The scrypt recipient type
+
+The scrypt recipient and identity implementations encrypt and decrypt the file
+key with a provided passphrase.
+
+#### scrypt recipient stanza
+
+An scrypt recipient stanza has three arguments.
+
+    -> scrypt ajMFur+EJLGaohv/dLRGnw 18
+    8SHBz/ldWnjyGFQqfjat6uNBarWqqEMDS7W8X7+Xq5Q
+
+The first is the string `scrypt`, the second is a base64-encoded salt computed
+by the recipient implementation as 16 bytes from a CSPRNG, and the third is the
+base-two logarithm of the scrypt work factor in decimal.
+
+A new salt MUST be generated for each stanza and each file.
+
+The body is computed as
+
+    wrap key = scrypt(N = work factor, r = 8, p = 1, dkLen = 32,
+        S = "age-encryption.org/v1/scrypt" || salt, P = passphrase)
+    body = ChaCha20-Poly1305(key = wrap key, plaintext = file key)
+
+where the ChaCha20-Poly1305 nonce is fixed as 12 0x00 bytes and scrypt is from
+[RFC 7914][].
+
+As long as any scrypt stanza is present, the identity implementation SHOULD
+reject any stanza that does not have `scrypt` as the first argument, and MUST
```

Comment:
`SHOULD` or `MUST`? Additionally, current clients require at most one `scrypt` stanza; this allows multiple.

Reply:
Good catch, redoing this to require it to be the only stanza.